pkg/p2p/message_handler_manager.go

```diff
@@ -0,0 +1,149 @@
+package p2p
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultHandlerOperationTimeout = 3 * time.Second
+)
+
+// MessageHandlerManager is for managing message topic handlers.
+// NOTE: for each topic, only one handler is allowed.
+type MessageHandlerManager interface {
+	RegisterHandler(ctx context.Context, topic Topic, tpi TypeInformation, fn HandlerFunc) (bool, error)
+
+	UnregisterHandler(ctx context.Context, topic Topic) (bool, error)
+
+	CheckError(ctx context.Context) error
+
+	// Clean unregisters all existing handlers.
+	Clean(ctx context.Context) error
+
+	// SetTimeout sets the timeout for handler operations.
+	// A timeout is needed because the underlying handler operations are
+	// asynchronous in the MessageServer.
+	SetTimeout(timeout time.Duration)
+}
+
+func newMessageHandlerManager(registrar handlerRegistrar) MessageHandlerManager {
+	return &messageHandlerManagerImpl{
+		messageServer: registrar,
+		timeout:       atomic.NewDuration(defaultHandlerOperationTimeout),
+		topics:        make(map[Topic]<-chan error),
+	}
+}
+
+// handlerRegistrar is an interface for the handler management-related
+// functionalities of MessageServer.
+// This interface is for easier unit-testing.
+type handlerRegistrar interface {
+	SyncAddHandler(context.Context, Topic, TypeInformation, HandlerFunc) (<-chan error, error)
+	SyncRemoveHandler(context.Context, Topic) error
+}
+
+type messageHandlerManagerImpl struct {
+	messageServer handlerRegistrar
+	// timeout is atomic to avoid unnecessary blocking
+	timeout *atomic.Duration
+
+	// mu protects topics
+	mu     sync.Mutex
+	topics map[Topic]<-chan error
+}
+
+func (m *messageHandlerManagerImpl) RegisterHandler(
+	ctx context.Context,
+	topic Topic,
+	tpi TypeInformation,
+	fn HandlerFunc,
+) (bool, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.topics[topic]; ok {
```

Comment:
Will it possible to replace the topic, or do we have to remove it at first before we replace it?

Reply:
We have to remove it as far as `MessageServer` is concerned. The implementation of the `MessageServer` does not support replacing it without removing it first. Can you think of a situation where exposing a `Replace` method is helpful?

Reply:
No, I can't figure out this senario either. That's fine